### DISCUSSION TITLE
Is it possible to index a vector-valued randomvar?

### DISCUSSION THREAD

**Discussion by mvsoom** [ReactiveBayes/RxInfer.jl#233] [Q&A] [answered]
Here I am again :)

I am trying to make an `AR(P)` model where the AR process acts upon a vector `h` which has a known MvNormal prior (given by the mean and covariance of a projected GP): `h ~ MvNormal(0, K)`.

I would like to use the `AR` node that comes with RxInfer rather than multiplying out covariance matrices as I suspect that this will be more stable/faster.

However, because `AR()` expects a state space vector of length `P` to act on, this would require me to index into `h` to yield appropriate slices of length `P` to feed into `AR()`. Is there a way round this?

---

If not, I would assign a MvNormal prior to the AR coefficients $a_{1:P}$ and embed them in the Toeplitz matrix $\Psi$ like this:
![image](https://github.com/ReactiveBayes/RxInfer.jl/assets/43237003/d3229585-eaeb-43af-9c72-477da107897f)
And the likelihood of the data $x$ would look like this:
![image](https://github.com/ReactiveBayes/RxInfer.jl/assets/43237003/c720d2f6-c1e0-444b-bdd5-7210c557abfb)
... where the kernel matrix $K$ is known and the unknowns are $a_{1:P},\nu_w,\nu_e$. Is this possible with RxInfer?

Thank you!

**Comment by albertpod** [comment, marked as answer]
Hi, @mvsoom. You have quite a peculiar likelihood. I'm unsure how to specify it yet, but let me answer your questions regardless of this likelihood.

So, first of all, what helps to understand what you can or cannot do within `RxInfer.jl` is to think of your model in terms of a graph. I know that this isn't something you want to dive in, and we try to hide this from the user so that you have a seamless experience.
So when you say `h[1] ` what does it mean in terms of the graph? Well, one way of taking the index is to do something like:
```julia
h ~ MvNormal(mean=m, cov=V)
h_1 ~ dot(h, c) # where c is a basis vector (1, 0, ..., 0)
```
So, by doing this, you can extract `h[1]` from `h.`

Another thing you can look at is the use of delta and/or ContinuousTransition nodes.

For example,
```julia
# say we have a function Psi_f(a) = [your matrix]
h ~ MvNormal(mean=m, cov=V)
Psi ~ Psi_f(h)
```

The problem is that your message from Psi is a matrix variate distribution. You'd most likely need to use CVI here, and I am not sure how `RxInfer.jl` will handle it (the current CVI implementation is not ideal; we are working on improving it).

If I were you, I would probably define a delta function of three arguments, i.e., `V ~ my_cov(a nu_1, nu_2)`, and pass the variable as a covariance of your likelihood. The backward message from your likelihood towards `V` is going to be Wishart, so in theory CVI should handle it. The rest shouldn't be a problem. 

@Nimrais have you tried something like that before?

**Reply by albertpod** [reply]
and @mvsoom, don't hesitate to ask questions and open more Q&A. We won't be able to reply fast due to the load of the team, but we will do that when we can.

**Reply by mvsoom** [reply]
I am looking into this; the first method works (the inference runs), but I need to do some more testing and compare to the second approach. Thank you!
